src/main/java/com/amazon/opendistroforelasticsearch/security/auditlog/impl/AuditMessageBuilder.java

```diff
@@ -0,0 +1,350 @@
+package com.amazon.opendistroforelasticsearch.security.auditlog.impl;
+
+import com.amazon.opendistroforelasticsearch.security.auditlog.AuditLog;
+import com.amazon.opendistroforelasticsearch.security.dlic.rest.support.Utils;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.rest.RestRequest;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class AuditMessageBuilder {
```

Comment:
If you're planning on using the Builder pattern to create AuditMessage objects, you need to implement this as a public static nested class of AuditMessage. See [link1](https://howtodoinjava.com/design-patterns/creational/builder-pattern-in-java/) [link2](https://dzone.com/articles/the-builder-pattern-for-class-with-many-constructo)

Reply:
This would make the interface to be 
`AuditMessage.AuditMessageBuilder()....` 

Why particular reason why this is needed ?

Reply:
Thats how the Builder pattern works. We just use Builders to construct other objects. We want an object of type AuditMessage, not of type AuditMessageBuilder. Whenever AuditMessage.Builder....build() is a common way to implement and use Builders. 

https://stackoverflow.com/questions/5007355/builder-pattern-in-effective-java

Reply:
Alright! Thanks for the links :)